core/cpuid.c

```diff
@@ -114,3 +131,234 @@ bool cpuid_host_has_feature_uncached(uint32_t feature_key)
+                                          uint32_t function, uint32_t index)
+{
+    int i;
+    hax_cpuid_entry *entry, *found = NULL;
+
+    for (i = 0; i < cpuid_info->total; ++i) {
+        entry = &cpuid_info->entries[i];
+        if (entry->function == function && entry->index == index) {
+            found = entry;
+            break;
+        }
+    }
+
+    return found;
+}
+
+static void cpuid_set_0000_0001(cpuid_t *cpuid, hax_cpuid *cpuid_info)
+{
+    const uint32_t kFunction = 0x01;
+    hax_cpuid_entry *entry;
+
+    entry = find_cpuid_entry(cpuid_info, kFunction, 0);
+    if (entry == NULL)
+        return;
+
+    hax_log(HAX_LOGI, "%s: function: %08lx, index: %lu, flags: %08lx\n",
+            __func__, entry->function, entry->index, entry->flags);
+    hax_log(HAX_LOGI, "%s: eax: %08lx, ebx: %08lx, ecx: %08lx, edx: %08lx\n",
+            __func__, entry->eax, entry->ebx, entry->ecx, entry->edx);
+
+    cpuid->feature_1_ecx = entry->ecx;
+    cpuid->feature_1_edx = entry->edx;
+
+    // Filter the unsupported features
+    cpuid->feature_1_ecx &= cache.host_supported.feature_1_ecx &
+                            cache.hax_supported.feature_1_ecx;
+    cpuid->feature_1_edx &= cache.host_supported.feature_1_edx &
+                            cache.hax_supported.feature_1_edx;
+
+    // Set fixed supported features
+    cpuid_set_fixed_features(cpuid);
+
+    if (entry->ecx != cpuid->feature_1_ecx ||
+        entry->edx != cpuid->feature_1_edx) {
+        hax_log(HAX_LOGW, "%s: filtered or unchanged flags: ecx: %08lx, "
+                "edx: %08lx\n", __func__, entry->ecx ^ cpuid->feature_1_ecx,
+                entry->edx ^ cpuid->feature_1_edx);
+    }
+}
+
+static void cpuid_set_8000_0001(cpuid_t *cpuid, hax_cpuid *cpuid_info)
+{
+    const uint32_t kFunction = 0x80000001;
+    hax_cpuid_entry *entry;
+
+    entry = find_cpuid_entry(cpuid_info, kFunction, 0);
+    if (entry == NULL)
+        return;
+
+    hax_log(HAX_LOGI, "%s: function: %08lx, index: %lu, flags: %08lx\n",
+            __func__, entry->function, entry->index, entry->flags);
+    hax_log(HAX_LOGI, "%s: eax: %08lx, ebx: %08lx, ecx: %08lx, edx: %08lx\n",
+            __func__, entry->eax, entry->ebx, entry->ecx, entry->edx);
+
+    cpuid->feature_8000_0001_edx = entry->edx;
+
+    // Filter the unsupported features
+    cpuid->feature_8000_0001_edx &=
+        cache.host_supported.feature_8000_0001_edx &
+        cache.hax_supported.feature_8000_0001_edx;
+
+    if (entry->edx != cpuid->feature_8000_0001_edx) {
+        hax_log(HAX_LOGW, "%s: filtered or unchanged flags: edx: %08lx\n",
+                __func__, entry->edx ^ cpuid->feature_8000_0001_edx);
+    }
+}
+
+static void cpuid_set_fixed_features(cpuid_t *cpuid)
+{
+    const uint32_t kFixedFeatures =
+        FEATURE(MCE)  |
+        FEATURE(APIC) |
+        FEATURE(MTRR) |
+        FEATURE(PAT);
+
+    cpuid->feature_1_edx |= kFixedFeatures;
```

Comment:
Silently adding flags to the ones provided by the user without even log this. I think when the user want to clear these bits this should be a error.

Reply:
Actually these flags were referred from your commits. As for logs, it will always output the unchanged bits as warning in the caller functions, such as cpuid_set_0000_0001().